### PR TITLE
NodeJS SDK required + optional types

### PR DESCRIPTION
**Pull request by dolanor**
Now we can use best of both world:
- required params are available directly in the function params
- optional params are available through an object that lets you mix and match the options you want

```diff
- withExec(args: string[], stdin?: string, redirectStdout?: string, redirectStderr?: string, experimentalPrivilegedNesting?: boolean): Container {
-
+ export type ContainerWithExecOpts = {
+   stdin?: string
+   redirectStdout?: string
+   redirectStderr?: string
+   experimentalPrivilegedNesting?: boolean
+ }
+ withExec(args: string[], opts?: ContainerWithExecOpts): Container {
```

### CODE REVIEW

**Comment by aluzzardi** [comment]
I believe we agreed to have objects only in Node.js, rather than mixing positional + objects 

https://discord.com/channels/707636530424053791/1040027260155088956/1043084201316077598

**Comment by dolanor** [comment]
> I believe we agreed to have objects only in Node.js, rather than mixing positional + objects
> 
> https://discord.com/channels/707636530424053791/1040027260155088956/1043084201316077598

I was basing myself on this, finally: https://github.com/dagger/dagger/issues/3810#issuecomment-1324034289
As it was better in a DX way, the LSP displaying each arg info popup directly when typing inside the function parenthesis.
Once in an `{}`, no popups shows. You have to explicitly open the completion with `CTRL+space`. I can see that it's not a crazy blocker either.
We just need to all be aligned.

- :x: flattened args, required or not. We all agree that it's not great. (current implementation)
- mix: required args are flattened, optionals are wrapped in a type (current PR)
- full object: required + optional types are wrapped in a type.

I lack the practice in TS to know which is more common (it would seem the full object? But then, I wonder why the DX is poorer by using this practice)

**Comment by slumbering** [comment]
> I believe we agreed to have objects only in Node.js, rather than mixing positional + objects
> 
> https://discord.com/channels/707636530424053791/1040027260155088956/1043084201316077598

We actually want to embrace both world. Like we said in Discord, full object is indeed very flexible. 
However the cons of that is IDE autocomplete doesn't give you any idea of what args is at least required. 

It's basically exactly like `Axios`
```ts
// url is required, but you can also pass optional config

axios.get(url[, config])

axios.get('/user', {
    params: {
      ID: 12345
    }
})
```

In our case
```ts
const image = await client
    .container()
    .from("alpine")
    .withExec(["apk", "add", "curl"], {
        experimentalPrivilegedNesting: true,
     })
```

**Comment by TomChv** [comment]
## TL;DR

I think **positional arguments + optional object** is a better solution for the current NodeJS SDK, but we shall not ignore the full object pattern because we may need to switch to it in the future (see **My Opinion section for details**).

## Reactions

> It would be exactly like Axios for instance

Other popular libraries like [Prisma](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#select-specific-fields) use full object patterns, both are valid and it mostly depends on the number of originally required parameters.

> However the cons of that is IDE autocomplete doesn't give you any idea of what args is at least required.

It does on Webstorm 

Screenshot 2022-11-30 at 16 05 52

And if not, you shall have an error on `build`, for example:

```shell
provisioning/docker-provision/image.ts:179:10 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'Test'.
  Property 'foo' is missing in type '{}' but required in type 'Test'.

179     test({})
             ~~

  provisioning/docker-provision/image.ts:172:7
    172       foo: string
              ~~~
    'foo' is declared here.


Found 1 error in provisioning/docker-provision/image.ts:179
```

## My opinion

### Overall

As long as operations have less than 3 required parameters and **N** number of optional, the better DX is to mix both positional arguments and objects for optional ones. This way, we keep a maximum of 4 arguments by operation.
More than 4 arguments are not acceptable... even 4 is already a lot IMO, but it shall still be possible to keep 80 characters
on a line.

However, if one operation does not fit these rules, we shall go for a full object pattern. 

### Default values problem

Another advantage of positional arguments is that we could have a default value. This does not make sense for our low-level operations, but I'm thinking about the future, and extensions may take advantage of this.
For instance 

```ts
function op(cmd: string, args: string[], magicNumber = 42, opts?: OptionalOpts) {...} 
```

The only problem with DX is that if you want to set `opts`, you will need to set `magicNumber` too.

This DX's problem is not possible with a full object pattern, but it's harder to define default values because you will need to also set values for every required parameter. At some point, this may lead to a bad DX with default parameters set to irrelevant values. 

For instance

```ts
interface Opt {
   cmd: string
   args: string[]
   magicNumber: number
   stdin?: string
   stdout?: string
}

// Set magicNumber to 42 by default but this require to also set
// cmd and args
const defaultOpt: Opt = {
   cmd: "", 
   args: [],
   magicNumber: 42
}

function op(opt: Opt = defaultOpt) { ... }
```

### DX comparison

The difference of DX is the following in the worst case (4 arguments)

```ts
// Fake operation arguments, for the example
interface OptionalOpArgs {
   stdin?: string
   stdout?: string
}

// Positional is simpler in that case and requires fewer lines to call it.
function positionalOp(cmd: string, args: string[], magicNumber: number, opts?: OptionalArgs) { ... }

positionalOp("foo", ["bar, "baz", "bool", "dog"], 42, {
  stdin: "standard"
})

// Full object is way more verbose but much clearer when you inline arguments directly with values.
interface OpArgs {
  cmd: string,
  args: string[],
  magicNumber: number,
  stdin?: string,
  stdout?: string
} 

function fullObjectOp(opts: OpAgs) { ... }

fullObjetOp({
  cmd: "foo",
  args: ["bar", "baz", "bool", "dog"],
  magicNumber: 42,
  stdin: "standard"
})
```

### Conclusion

I compared the pros and cons of both solutions before, my final opinion is that we should choose **positional arguments + optional objects for now**.

I went through [`client.gen.ts`](https://github.com/dagger/dagger/blob/67cadb8e425dea49a2209e2133e432a4464d1702/sdk/nodejs/api/client.gen.ts) of this PR, and the maximum number of arguments I've seen is 2, so there are no red flags about adopting this solution.
On top of that, this follows the current Go implementation, so we keep a consistent DX across all SDKs. 
Indeed, we shall take advantage of Typescript default values if we can (maybe extract those from comments?), but the more we stick to one common DX, the better.

With the evolution of Dagger, this implementation may change to a full object, especially if users create extensions with more than three required arguments. However, they could add those arguments to the 4th object and make it required. This is not the best solution, but it's probably the best trade for now if this situation happens.

The main issue with the full object pattern is the limitation of default values which may become a problem in the future. We need to find a solution before adopting this pattern (if we have to) one day.

**Comment by dolanor** [comment]
Nice and well exposed roundup.
Out of it, it seems this current approach is good, but it seems the full object is more future proof.
Especially when extensions will get out, we will lose control on the extension API, and with that, the number of argument people will use. But maybe it's their problem, then.
And maybe we can readjust when it happens?

**Comment by slumbering** [comment]
Thanks @TomChv for the clarification.

> Other popular libraries like [Prisma](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#select-specific-fields) use full object patterns, both are valid and it mostly depends on the number of originally required parameters.

True. But many other libraries use both required + object, Axios, Cypress, GraphQLClient from Graphql-request... What I mean is what is the best feat for us and what are we looking for.
In your example you are showing typescript error in Webstorm (which I also get on VSCode). What I wanted to highlight is the autocomplete:

Capture d’écran 2022-12-01 à 09 26 08

Ok, what is Test, I need to browser the SDK to see what is required and what is not.

Versus:

Capture d’écran 2022-12-01 à 09 44 13

Oh right, the url is required.

> This DX's problem is not possible with a full object pattern, but it's harder to define default values because you will need to also set values for every required parameter. At some point, this may lead to a bad DX with default parameters set to irrelevant values.

But can't you just pick the value you want without the need of settings every other values ?

```ts
interface Opt {
  cmd: string;
  args: string[];
  magicNumber: number;
  stdin?: string;
  stdout?: string;
}

// Set magicNumber to 42 by default but this require to also set
// cmd and args
const defaultOpt: Pick<Opt, "magicNumber"> = {
  magicNumber: 42,
};
```

**Comment by slumbering** [comment]
> Nice and well exposed roundup. Out of it, it seems this current approach is good, but it seems the full object is more future proof. Especially when extensions will get out, we will lose control on the extension API, and with that, the number of argument people will use. But maybe it's their problem, then. And maybe we can readjust when it happens?

I would argue that this problem cover every SDK. We are proposing a way of writing pipeline. I guess we will provide a tutorial of how to write an extension.

**Comment by helderco** [comment]
In Python optional values are the ones with a default. So if the api has a default in a param just put it in `opts`, don’t make it positional.

**Comment by dolanor** [comment]
Yes, for me, optional, or with a default value has the same semantic for me. You can optionally pass a value.
If not, a default value will be used: the zero value for the language, or a default passed in the signature
But, all in all, it is optional. I would put it into the options.

**Comment by dolanor** [comment]
I tried to rebase on `main`, but got blocked when I resolved some conflicts. The spec test didn't pass:
```console
#13 docker-entrypoint.sh yarn test
#13 0.433 yarn run v1.22.19
#13 0.482 $ mocha
#13 0.718 (node:40) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
#13 0.718 (Use `node --trace-warnings ...` to show where the warning was created)
#13 4.195 
#13 4.196 
#13 4.196   NodeJS SDK api
#13 4.199     ✔ Build correctly a query with one argument
#13 4.200     ✔ Build correctly a query with different args type
#13 4.201     ✔ Build one query with multiple arguments
#13 4.202     ✔ Build a query by splitting it
#13 4.221     ✔ Pass a client with an implicit ID as a parameter
#13 4.222     ✔ Build a query with positionnal and optionals arguments
#13 4.222     ✔ Test Field Immutability
#13 4.224     ✔ Return a flatten Graphql response
#13 4.225     ✔ Return a error for Graphql object nested response
#13 4.225 
#13 4.225   Provisioner
#13 4.225     Docker Image
#13 4.252       1) Should create a GQL client on call to connect
#13 4.253 
#13 4.253   NodeJS sdk Connect
#13 6.774     ✔ Connect to local engine and execute a simple query to make sure it does not fail (2521ms)
#13 6.777 
#13 6.778 
#13 6.779   10 passing (3s)
#13 6.779   1 failing
#13 6.779 
#13 6.780   1) Provisioner
#13 6.780        Docker Image
#13 6.780          Should create a GQL client on call to connect:
#13 6.780      Error: failed to copy engine session binary: Error: Command failed with ENOENT: docker run --rm --entrypoint /bin/cat ghcr.io/dagger/engine:v0.3.6@sha256:d124de285773409fd83d203d3f46364e20149ee46842c50eec6cc51c50fffb62 /usr/bin/dagger-engine-session-linux-amd64
#13 6.780 spawnSync docker ENOENT
#13 6.780       at DockerImage.pullEngineSessionBin (file:///workdir/provisioning/docker-provision/image.ts:199:13)
#13 6.780       at DockerImage.<anonymous> (file:///workdir/provisioning/docker-provision/image.ts:96:12)
#13 6.780       at Generator.next (<anonymous>)
#13 6.780       at file:///workdir/provisioning/docker-provision/image.ts:7:71
#13 6.780       at new Promise (<anonymous>)
#13 6.780       at __awaiter (file:///workdir/provisioning/docker-provision/image.ts:3:12)
#13 6.780       at DockerImage.Connect (file:///workdir/provisioning/docker-provision/image.ts:79:16)
#13 6.780       at Context.<anonymous> (file:///workdir/provisioning/test/provisioner.spec.ts:15:40)
#13 6.780       at Generator.next (<anonymous>)
#13 6.780       at file:///workdir/provisioning/test/provisioner.spec.ts:7:71
#13 6.780       at new Promise (<anonymous>)
#13 6.780       at __awaiter (file:///workdir/provisioning/test/provisioner.spec.ts:3:12)
#13 6.780       at Context.<anonymous> (file:///workdir/provisioning/test/provisioner.spec.ts:16:20)
#13 6.780       at processImmediate (node:internal/timers:466:21)
#13 6.780 
#13 6.780 
#13 6.780 
#13 6.830 error Command failed with exit code 1.
#13 6.830 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: input:1: container.from.withWorkdir.withRootfs.withExec.withRootfs.withMountedDirectory.withExec.exitCode process "docker-entrypoint.sh yarn test" did not complete successfully: exit code: 1

Please visit https://dagger.io/help#go for troubleshooting guidance.
exit status 1
```
Running the command locally works
`docker run --rm --entrypoint /bin/cat ghcr.io/dagger/engine:v0.3.6@sha256:d124de285773409fd83d203d3f46364e20149ee46842c50eec6cc51c50fffb62 /usr/bin/dagger-engine-session-linux-amd64` pulls an image and then output the binary in the stdout.
I saved it to my FS and get
```
dolanor@eastpeak:~/src/dagger$ docker run --rm --entrypoint /bin/cat ghcr.io/dagger/engine:v0.3.6@sha256:d124de285773409fd83d203d3f46364e20149ee46842c50eec6cc51c50fffb62 /usr/bin/dagger-engine-session-linux-amd64 > output
dolanor@eastpeak:~/src/dagger$ file output 
output: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=BjqPKTVnK5mzIxipDFhU/P4-IJcmun0XTdRUzYTR7/pHcRvJ-_G1bUgrcpMjAt/ZYmvJL30KLsEuEqjivz9, stripped
dolanor@eastpeak:~/src/dagger$ GOWORK=off go version -m output  
output: go1.19.3
	path	github.com/dagger/dagger/cmd/engine-session
	mod	github.com/dagger/dagger	(devel)	
	dep	github.com/agext/levenshtein	v1.2.3	h1:YB2fHEn0UJagG8T1rrWknE3ZQzWM06O8AMAatNn7lmo=
	dep	github.com/cenkalti/backoff/v4	v4.1.3	h1:cFAlzYUlVYDysBEH2T5hyJZMh3+5+WCBvSnK6Q8UtC4=
	dep	github.com/containerd/console	v1.0.3	h1:lIr7SlA5PxZyMV30bDW0MGbiOPXwc63yRuCP0ARubLw=
	dep	github.com/containerd/containerd	v1.6.10	h1:8aiav7I2ZyQLbTlNMcBXyAU1FtFvp6VuyuW13qSd6Hk=
	dep	github.com/containerd/continuity	v0.3.0	h1:nisirsYROK15TAMVukJOUyGJjz4BNQJBVsNvAXZJ/eg=
	dep	github.com/containerd/ttrpc	v1.1.0	h1:GbtyLRxb0gOLR0TYQWt3O6B0NvT8tMdorEHqIQo/lWI=
	dep	github.com/containerd/typeurl	v1.0.2	h1:Chlt8zIieDbzQFzXzAeBEF92KhExuE4p9p92/QmY7aY=
	dep	github.com/dagger/graphql	v0.0.0-20221102000338-24d5e47d3b72	h1:0I9Y9nsFVcP42k9eOpjYmcm+NMEIFxNuJFOHfQ8MKpo=
	dep	github.com/dagger/graphql-go-tools	v0.0.0-20221102001222-e68b44170936	h1:yRwbZ5iT34iXf2Kg3my9Yclbg5Mq20b+w+uqePfbUoo=
	dep	github.com/docker/cli	v20.10.17+incompatible	h1:eO2KS7ZFeov5UJeaDmIs1NFEDRf32PaqRpvoEkKBy5M=
	dep	github.com/docker/distribution	v2.8.1+incompatible	h1:Q50tZOPR6T/hjNsyc9g8/syEs6bk8XXApsHjKukMl68=
	dep	github.com/docker/docker	v20.10.17+incompatible
	=>	github.com/docker/docker	v20.10.3-0.20220414164044-61404de7df1a+incompatible	h1:jnhU41Zm9biz6e4GPOvAtuognZ+pi8S77ZlRir/SHRA=
	
	dep	github.com/docker/docker-credential-helpers	v0.6.4	h1:axCks+yV+2MR3/kZhAmy07yC56WZ2Pwu/fKWtKuZB0o=
	dep	github.com/docker/go-connections	v0.4.0	h1:El9xVISelRB7BuFusrZozjnkIM5YnzCViNKohAFqRJQ=
	dep	github.com/docker/go-units	v0.4.0	h1:3uh0PgVws3nIA0Q+MwDC8yjEPf9zjRfZZWXZYDct3Tw=
	dep	github.com/go-logr/logr	v1.2.3	h1:2DntVwHkVopvECVRSlL5PSo9eG+cAkDCuckLubN+rq0=
	dep	github.com/go-logr/stdr	v1.2.2	h1:hSWxHoqTgW2S2qGc0LTAI563KZ5YKYRhT3MFKZMbjag=
	dep	github.com/gofrs/flock	v0.8.1	h1:+gYjHKf32LDeiEEFhQaotPbLuUXjY5ZqxKgXy7n59aw=
	dep	github.com/gogo/googleapis	v1.4.1	h1:1Yx4Myt7BxzvUr5ldGSbwYiZG6t9wGBZ+8/fX3Wvtq0=
	dep	github.com/gogo/protobuf	v1.3.2	h1:Ov1cvc58UF3b5XjBnZv7+opcTcQFZebYjWzi34vdm4Q=
	dep	github.com/golang/protobuf	v1.5.2	h1:ROPKBNFfQgOUMifHyP+KYbvpjbdoFNs+aK7DXlji0Tw=
	dep	github.com/google/shlex	v0.0.0-20191202100458-e7afc7fbc510	h1:El6M4kTTCOh6aBiKaUGG7oYTSPP8MxqL4YI3kZKwcP4=
	dep	github.com/grpc-ecosystem/go-grpc-middleware	v1.3.0	h1:+9834+KizmvFV7pXQGSXQTsaWhq2GjuNUt0aUU0YBYw=
	dep	github.com/grpc-ecosystem/grpc-gateway/v2	v2.11.2	h1:BqHID5W5qnMkug0Z8UmL8tN0gAy4jQ+B4WFt8cCgluU=
	dep	github.com/klauspost/compress	v1.15.9	h1:wKRjX6JRtDdrE9qwa4b/Cip7ACOshUI4smpCQanqjSY=
	dep	github.com/moby/buildkit	v0.10.6	h1:DJlEuLIgnu34HQKF4n9Eg6q2YqQVC0eOpMb4p2eRS2w=
	dep	github.com/moby/locker	v1.0.1	h1:fOXqR41zeveg4fFODix+1Ch4mj/gT0NE1XJbp/epuBg=
	dep	github.com/moby/sys/signal	v0.7.0	h1:25RW3d5TnQEoKvRbEKUGay6DCQ46IxAVTT9CUMgmsSI=
	dep	github.com/morikuni/aec	v1.0.0	h1:nP9CBfwrvYnBRgY6qfDQkygYDmYwOilePFkwzv4dU8A=
	dep	github.com/opencontainers/go-digest	v1.0.0	h1:apOUWs51W5PlhuyGyz9FCeeBIOUDA/6nW8Oi/yOhh5U=
	dep	github.com/opencontainers/image-spec	v1.0.3-0.20211202183452-c5a74bcca799	h1:rc3tiVYb5z54aKaDfakKn0dDjIyPpTtszkjuMzyt7ec=
	dep	github.com/pkg/errors	v0.9.1	h1:FEBLx1zS214owpjy7qsBeixbURkuhQAwrK5UwLGTwt4=
	dep	github.com/sirupsen/logrus	v1.9.0	h1:trlNQbNUG3OdDrDil03MCb1H2o9nJ1x4/5LYw7byDE0=
	dep	github.com/spf13/cobra	v1.6.1	h1:o94oiPyS4KD1mPy2fmcYYHHfCxLqYjJOhGsCHFZtEzA=
	dep	github.com/spf13/pflag	v1.0.5	h1:iy+VFUOCP1a+8yFto/drg2CJ5u0yRoB7fZw3DKv/JXA=
	dep	github.com/tonistiigi/fsutil	v0.0.0-20220315205639-9ed612626da3	h1:T1pEe+WB3SCPVAfVquvfPfagKZU2Z8c1OP3SuGB+id0=
	dep	github.com/tonistiigi/units	v0.0.0-20180711220420-6950e57a87ea	h1:SXhTLE6pb6eld/v/cCndK0AMpt1wiVFb/YYmqB3/QG0=
	dep	github.com/tonistiigi/vt100	v0.0.0-20210615222946-8066bb97264f	h1:DLpt6B5oaaS8jyXHa9VA4rrZloBVPVXeCtrOsrFauxc=
	dep	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc	v0.34.0	h1:PNEMW4EvpNQ7SuoPFNkvbZqi1STkTPKq+8vfoMl/6AE=
	dep	go.opentelemetry.io/otel	v1.10.0	h1:Y7DTJMR6zs1xkS/upamJYk0SxxN4C9AqRd77jmZnyY4=
	dep	go.opentelemetry.io/otel/exporters/jaeger	v1.4.1	h1:VHCK+2yTZDqDaVXj7JH2Z/khptuydo6C0ttBh2bxAbc=
	dep	go.opentelemetry.io/otel/exporters/otlp/internal/retry	v1.9.0	h1:ggqApEjDKczicksfvZUCxuvoyDmR6Sbm56LwiK8DVR0=
	dep	go.opentelemetry.io/otel/exporters/otlp/otlptrace	v1.9.0	h1:NN90Cuna0CnBg8YNu1Q0V35i2E8LDByFOwHRCq/ZP9I=
	dep	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc	v1.9.0	h1:M0/hqGuJBLeIEu20f89H74RGtqV2dn+SFWEz9ATAAwY=
	dep	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp	v1.9.0	h1:FAF9l8Wjxi9Ad2k/vLTfHZyzXYX72C62wBGpV3G6AIo=
	dep	go.opentelemetry.io/otel/sdk	v1.9.0	h1:LNXp1vrr83fNXTHgU8eO89mhzxb/bbWAsHG6fNf3qWo=
	dep	go.opentelemetry.io/otel/trace	v1.10.0	h1:npQMbR8o7mum8uF95yFbOEJffhs1sbCOfDh8zAJiH5E=
	dep	go.opentelemetry.io/proto/otlp	v0.18.0	h1:W5hyXNComRa23tGpKwG+FRAc4rfF6ZUg1JReK+QHS80=
	dep	golang.org/x/crypto	v0.0.0-20220722155217-630584e8d5aa	h1:zuSxTR4o9y82ebqCUJYNGJbGPo6sKVl54f/TVDObg1c=
	dep	golang.org/x/net	v0.0.0-20220811182439-13a9a731de15	h1:cik0bxZUSJVDyaHf1hZPSDsU8SZHGQZQMeueXCE7yBQ=
	dep	golang.org/x/sync	v0.0.0-20220722155255-886fb9371eb4	h1:uVc8UZUe6tr40fFVnUP5Oj+veunVezqYl9z7DYw9xzw=
	dep	golang.org/x/sys	v0.0.0-20220811171246-fbc7d0a398ab	h1:2QkjZIsXupsJbJIdSjjUOgWK3aEtzyuh2mPt3l/CkeU=
	dep	golang.org/x/text	v0.4.0	h1:BrVqGRd7+k1DiOgtnFvAkoQEWQvBc25ouMJM6429SFg=
	dep	golang.org/x/time	v0.0.0-20220722155302-e5dcc9cfc0b9	h1:ftMN5LMiBFjbzleLqtoBZk7KdJwhuybIU+FckUHgoyQ=
	dep	google.golang.org/genproto	v0.0.0-20220810155839-1856144b1d9c	h1:IooGDWedfLC6KLczH/uduUsKQP42ZZYhKx+zd50L1Sk=
	dep	google.golang.org/grpc	v1.51.0	h1:E1eGv1FTqoLIdnBCZufiSHgKjlqG6fKFf6pPWtMTh8U=
	dep	google.golang.org/protobuf	v1.28.1	h1:d0NfwRgPtno5B1Wa6L2DAG+KivqkdutMf1UhdNx175w=
	build	-compiler=gc
	build	-ldflags="-s -w"
	build	CGO_ENABLED=0
	build	GOARCH=amd64
	build	GOOS=linux
	build	GOAMD64=v1
	build	vcs=git
	build	vcs.revision=b1b8dff3c3aebad58a4ed21c94e25215831ae08c
	build	vcs.time=2022-11-30T17:54:41Z
	build	vcs.modified=true
```

have anyone encountered that?
cc @sipsma @vito @aluzzardi @slumbering

File: codegen/generator/nodejs/templates/functions.go [hunk 1]
```diff
@@ -1,22 +1,44 @@
...
+	"dagger.io/dagger"
 	"github.com/iancoleman/strcase"
 
+	"github.com/dagger/dagger/codegen/generator"
 	"github.com/dagger/dagger/codegen/introspection"
 )
 
+var currentSchema *introspection.Schema
+
+func init() {
```
Comment (aluzzardi) [inline codegen/generator/nodejs/templates/functions.go:18]: We shouldn't use `dagger.Connect()` in an `init()` function

**Review by aluzzardi — CHANGES_REQUESTED** [review]

**Comment by aluzzardi** [comment]
> We actually want to embrace both world. Like we said in Discord, full object is indeed very flexible.
> However the cons of that is IDE autocomplete doesn't give you any idea of what args is at least required.

SGTM.

Whatever makes most sense in TS/JS. The goal is to have the most native framework, not to copy what Go or Python are doing.

The approach was adopted in Go for very Go specific reasons: it's annoying to use structures in Go e.g. `c.WithExec(dagger.ContainerWithExecOpts{Args: []string{"ls"}})` vs `c.withExec({args: ["ls"])` in TS

**Comment by aluzzardi** [comment]
@dolanor Please squash commits before merge. PRs should have 2-3 commits at most

File: codegen/generator/nodejs/templates/functions.go (hunk 1 above)
Comment (dolanor) [reply codegen/generator/nodejs/templates/functions.go:18]: Yep, this is an error, it was supposed to be there for my tests. I'm gonna move it into the tests.

**Comment by dolanor** [comment]
Will squash.

File: codegen/generator/nodejs/templates/functions.go (hunk 1 above)
Comment (dolanor) [reply codegen/generator/nodejs/templates/functions.go]: fixed.

**Comment by dolanor** [comment]
Added the commits from #4031 since we're less in an urge to release now that v0.2.0 has been released.

**Comment by dolanor** [comment]
Somehow, the connect method signature in the docs is:

**Review by TomChv — APPROVED** [review]
LGTM, the DX looks nice!

**Comment by gerhard** [comment]
We cannot merge this @aluzzardi until you re-review it. All yours now 💪

**Review by aluzzardi — APPROVED** [review]

**Comment by aluzzardi** [comment]
LGTM -- we should consider doing a release once this is in

**Comment by dolanor** [comment]
Doing the rebase right now.

**Comment by gerhard** [comment]
Well done all for pushing this through 💪
